torch/csrc/jit/serialization/export_module.cpp

```diff
@@ -134,38 +159,55 @@ c10::IValue getFunctionTuple(const Function& func) {
                       {"types", Tup(types)},
                       {"register_size", register_size}});
 
-  return Tup({func.qualname().qualifiedName(), table});
+  if (!save_debug_info) {
+    return Tup({func.qualname().qualifiedName(), table});
+  } else {
+    // module debug info
+    std::vector<IValue> module_paths;
+    module_paths.reserve(op_module_paths.size());
+    for (auto& path : op_module_paths) {
+      module_paths.emplace_back(std::move(path));
+    }
+
+    auto debug_info = Table({{"module_debug_info", Tup(module_paths)}});
+    return Tup({func.qualname().qualifiedName(), table, debug_info});
+  }
 }
 
-void setstateTuple(const IValue& ivalue, std::vector<c10::IValue>& elements) {
+void setstateTuple(
+    const Module& module,
+    const IValue& ivalue,
+    std::vector<c10::IValue>& elements,
+    bool save_debug_info) {
   if (!ivalue.isObject())
     return;
   auto obj = ivalue.toObject();
   auto type = obj->type();
   if (checkHasValidSetGetState(type)) {
     Function& setstate = type->getMethod("__setstate__");
     if (setstate.isGraphFunction()) {
-      elements.push_back(getFunctionTuple(setstate));
+      elements.push_back(getFunctionTuple(module, setstate, save_debug_info));
     }
   } else {
     for (size_t i = 0, n = type->numAttributes(); i < n; ++i) {
-      setstateTuple(obj->getSlot(i), elements);
+      setstateTuple(module, obj->getSlot(i), elements, save_debug_info);
     }
   }
 }
 } // namespace
 
 void moduleMethodsTuple(
     const Module& module,
-    std::vector<c10::IValue>& elements) {
+    std::vector<c10::IValue>& elements,
+    bool save_debug_info) {
   auto methods = module.get_methods();
   // top level methods
   for (const auto& method : methods) {
-    elements.push_back(getFunctionTuple(method.function()));
+    elements.push_back(getFunctionTuple(module, method.function(), save_debug_info));
   }
 
   // __setstate__ of all components
-  setstateTuple(module._ivalue(), elements);
+  setstateTuple(module, module._ivalue(), elements, save_debug_info);
```

Comment:
I am wondering if we should do `    ReconstructScopes(module, *graph, "top");` here and avoid passing module around. I think it makes sense to do it here. It is effectively saying, let's populate some debug info. And I think it does not need to be done closest to where it will be saved.

Reply:
I thought about it yesterday. I'm not sure whether we can call `ReconstructScopes` here, because the `graph` did change each time we call the function. However, I was thinking that we can perhaps move `constructFunctionToModuleMap` to this file and only pass that map to `ReconstructScopes` (because our `module` stay unchanged, so it doesn't make sense to construct that map again and again in `ReconstructScopes`. What do you think?

Reply:
Aah nevermind. Yeah we need to do this after inlining and not before. Let's see.